warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/facet/FacetHandler.java

```diff
@@ -0,0 +1,363 @@
+    }
+    
+    @Override
+    public IngestHelperInterface getHelper(Type datatype) {
+        return datatype.getIngestHelper(this.taskAttemptContext.getConfiguration());
+    }
+    
+    @Override
+    public void close(TaskAttemptContext context) {
+        /* no-op */
+    }
+    
+    @Override
+    public RawRecordMetadata getMetadata() {
+        return null;
+    }
+    
+    protected byte[] flatten(ColumnVisibility vis) {
+        return markingFunctions == null ? vis.flatten() : markingFunctions.flatten(vis);
+    }
+    
+    @Override
+    public long process(KEYIN key, RawRecordContainer event, Multimap<String,NormalizedContentInterface> fields,
+                    TaskInputOutputContext<KEYIN,? extends RawRecordContainer,KEYOUT,VALUEOUT> context, ContextWriter<KEYOUT,VALUEOUT> contextWriter)
+                    throws IOException, InterruptedException {
+        
+        final String shardId = shardIdFactory.getShardId(event).substring(0, 8);
+        Text dateColumnQualifier = new Text(shardId); // TODO: Makes an assumption about the structure of the shardId.
+        
+        HyperLogLogPlus cardinality = new HyperLogLogPlus(10);
+        cardinality.offer(event.getId().toString());
+        
+        Text cv = new Text(flatten(event.getVisibility()));
+        
+        final HashTableFunction<KEYIN,KEYOUT,VALUEOUT> func = new HashTableFunction<>(contextWriter, context, facetHashTableName, facetHashThreshold,
+                        event.getDate());
+        Multimap<String,NormalizedContentInterface> eventFields = hashEventFields(fields, func);
+        
+        Stream<String> eventFieldKeyStream = eventFields.keySet().stream().filter(new TokenPredicate());
+        if (fieldFilter != null) {
+            eventFieldKeyStream = eventFieldKeyStream.filter(fieldFilter);
+        }
+        Set<String> keySet = eventFieldKeyStream.collect(Collectors.toSet());
+        List<Set<String>> keySetList = Lists.newArrayList();
+        keySetList.add(keySet);
+        keySetList.add(keySet);
+        
+        long countWritten = 0;
+        
+        Value sharedValue = new Value(cardinality.getBytes());
+        Multimap<BulkIngestKey,Value> results = ArrayListMultimap.create();
+        
+        for (String pivotFieldName : pivotMap.keySet()) {
+            Text reflexiveCf = createColumnFamily(pivotFieldName, pivotFieldName);
+            for (NormalizedContentInterface pivotTypes : eventFields.get(pivotFieldName)) {
+                for (String facetFieldName : pivotMap.get(pivotFieldName)) {
+                    if (pivotFieldName.equals(facetFieldName))
+                        continue;
+                    
+                    Text generatedCf = createColumnFamily(pivotFieldName, facetFieldName);
+                    Text myCf = generatedCf;
+                    
+                    if (HashTableFunction.isReduced(pivotTypes))
```

Comment:
Shouldn't this check be moved outside of this for loop but inside the containing loop (i.e move to line 247)?  Alternatively the eventFields could be prefiltered to not include any "reduced" members.

Reply:
Yes `HashTableFunction.isReduced(pivotTypes)` can be moved out. 

For what it's worth, we do use the reduced members in `eventFields` for the inner loop that iterates over `facetFieldName` members when generating facets, so I can't remove all reduced members from `eventFields` (see snippet below). I suppose I could create separate collections of pivot and facet fields, but is that any better?

```
for (NormalizedContentInterface facetTypes : eventFields.get(facetFieldName)) {
                        if (HashTableFunction.isReduced(facetTypes)) {
                            myCf.append(HashTableFunction.FIELD_APPEND_BYTES, 0, HashTableFunction.FIELD_APPEND_BYTES.length);
                        }
```

Reply:
your choice.

Reply:
Done. I just moved `HashTableFunction.isReduced(pivotTypes)` outside of the loop it was in now.